README.md

[![NPM version][npm-version-image]][npm-url]
[![NPM downloads][npm-downloads-image]][npm-url]
[![MIT License][license-image]][license-url]
[![js-standard-style][standard-style-image]][standard-style-url]
[![Build Status: Linux][travis-image]][travis-url]
[![Build Status: Windows][appveyor-image]][appveyor-url]
[![Coverage Status][coveralls-image]][coveralls-url]


# create-raml

### Create RAML from object or Express.js application

 v3.2.4


## Dependencies

- Node.js v.6.0 or higher


## Installation
```npm i -S create-raml```


## Create RAML based on Express.js

### Simple express example

```javascript
var express = require('express');
var Raml = require('create-raml');

var app = express();

var raml = new Raml({ express: app });

// regular app express workflow ( app.get, app.post, app.listen... etc )
```

### Get created raml

```curl 127.0.0.1:3000/api.raml```


### Extended express example

- [extended Express API example](docs/express_movies_api.md) - movies database API example ( GET, POST, DELETE methods; RAM data storage ). Result example as html: [Movies Database API documentation](http://create-raml.simpleness.org/express_movies_api.html)

### Simple example

```javascript
var express = require('express');
var Raml = require('create-raml');

var app = express();
var raml = new Raml({ express: app });

app.get('/movies', function (req, res) { res.send('List of all movies'); });
app.post('/movies', function (req, res) { res.send('Add new movie'); });
app.get('/movies/:id', function (req, res) { res.send('Get movie by id'); });
app.delete('/movies/:id', function (req, res) { res.send('Delete movie by id'); });

app.listen(3000, function () { console.log('Example app listening on port 3000!'); });
```

```curl 127.0.0.1:3000/api.raml```

### Result

```
#%RAML 1.0
title: 
version: 

types:

/api.raml:
  get:
    description: get /api.raml

/movies:
  get:
    description: get /movies
  post:
    description: post /movies
    
  /{id}:
    get:
      description: get /movies/:id
    delete:
      description: delete /movies/:id
```

## Create RAML from object

```javascript
var Raml = require('create-raml');
var raml = new Raml({
  title: 'Testing',
  baseUri: 'http://localhost:3000',
  version: 'v1',
});

raml.type('books', {
  name: { type: 'string', required: true },
  numberOfPages: { type: 'integer' },
});

raml.methods('books', 'get', {
  description: 'Get information about all books',
  responses: {
    200: { 'application/json': [{ name: 'one', author: { name: 'Art' } }] },
    404: { 'application/json': { code: '120', message: 'Books not found' } },
  },
});

raml.generate(function (err, ramlText) {
  console.log(ramlText);
});
```


### Result

```
#%RAML 1.0
title: Testing
baseUri: http://localhost:3000
version: v1

types:
  books: |
     {
       "name": {
         "type": "string",
         "required": true
       },
       "numberOfPages": {
         "type": "integer"
       }
     }

/books:
  get:
    description: Get information about all books
    responses:
      200:
        body:
          application/json:
            example: |
             [
               {
                 "name": "one",
                 "author": {
                   "name": "Art"
                 }
               }
             ]
      404:
        body:
          application/json:
            example: |
             {
               "code": "120",
               "message": "Books not found"
             }
```

## Options parameters

```javascript
var raml = new Raml(options);
```

 - **version** - version of RAML ( default: 1.0 )
 - **express** - an Express application
 - **path** - path to get API RAML ( default: /api.raml )
 - **storeResponses** - store first response as example ( default: false )
 - **guessAll** - make description quite pretty ( default: false )
 - **title** - title of API in document
 - **baseUri** - URI of API in document
 - **versionAPI** - version of API in document
 - **templateFileName** - path to template


## Tests

```npm test```


## Change Log

[all changes](CHANGELOG.md)


## Created by

Dimitry, 2@ivanoff.org.ua

```curl -A cv ivanoff.org.ua```


[license-image]: http://img.shields.io/badge/license-MIT-blue.svg?style=flat
[license-url]: LICENSE

[standard-style-image]: https://img.shields.io/badge/code%20style-airbnb-blue.svg?style=flat
[standard-style-url]: https://github.com/airbnb/javascript

[npm-url]: https://npmjs.org/package/create-raml
[npm-version-image]: http://img.shields.io/npm/v/create-raml.svg?style=flat
[npm-downloads-image]: http://img.shields.io/npm/dm/create-raml.svg?style=flat

[travis-url]: https://travis-ci.org/ivanoff/create-raml
[travis-image]: https://travis-ci.org/ivanoff/create-raml.svg?branch=master

[appveyor-url]: https://ci.appveyor.com/project/ivanoff/create-raml/branch/master
[appveyor-image]: https://ci.appveyor.com/api/projects/status/lp3nhnam1eyyqh33/branch/master?svg=true

[coveralls-url]: https://coveralls.io/github/ivanoff/create-raml
[coveralls-image]: https://coveralls.io/repos/github/ivanoff/create-raml/badge.svg


CHANGELOG.md
# create-raml changelog

## [3.2.4](http://github.com/ivanoff/create-raml/tree/3.2.4) (2018-10-18)
[Full Changelog](http://github.com/ivanoff/create-raml/compare/3.2.3...3.2.4)

**What Was Done:**

- fix dependencies


## [3.2.3](http://github.com/ivanoff/create-raml/tree/3.2.3) (2017-03-23)
[Full Changelog](http://github.com/ivanoff/create-raml/compare/3.2.2...3.2.3)

**What Was Done:**

- add simple express example


## [3.2.2](http://github.com/ivanoff/create-raml/tree/3.2.2) (2017-03-07)
[Full Changelog](http://github.com/ivanoff/create-raml/compare/3.1.1...3.2.2)

**What Was Done:**

- guessAll implemented to make description quite pretty
- describe options parameters


## [3.1.1](http://github.com/ivanoff/create-raml/tree/3.1.1) (2017-03-07)
[Full Changelog](http://github.com/ivanoff/create-raml/compare/3.0.1...3.1.1)

**What Was Done:**

- fixing Unknown node error


## [3.0.1](http://github.com/ivanoff/create-raml/tree/3.0.1) (2017-03-05)
[Full Changelog](http://github.com/ivanoff/create-raml/compare/2.0.1...3.0.1)

**What Was Done:**

- add express support


## [2.0.1](http://github.com/ivanoff/create-raml/tree/2.0.1) (2017-03-01)
[Full Changelog](http://github.com/ivanoff/create-raml/compare/1.1.1...2.0.1)

**What Was Done:**

- add raml2obj validator test
- add RAML v0.8


## [1.1.1](http://github.com/ivanoff/create-raml/tree/1.1.1) (2017-02-28)
[Full Changelog](http://github.com/ivanoff/create-raml/compare/1.0.1...1.1.1)

**What Was Done:**

- make all tests
- make short readme


## [1.0.1](http://github.com/ivanoff/create-raml/tree/1.0.1) (2017-02-10)
[Full Changelog](http://github.com/ivanoff/create-raml/compare/1.0.1...1.0.1)

**What Was Done:**

- make first run


test/index.js
'use strict';
var should = require('chai').should();
var Raml = require('../');

describe('testing v1.0 with no validators', function () {

    describe('check if text is in generated raml', function () {

        beforeEach(function () {
          this.raml = new Raml({
            title: 'Testing',
            baseUri: 'http://localhost:3000',
            versionAPI: 'v1',
          });
          this.raml.type('books', {
            name: { type: 'string', required: true },
            numberOfPages: { type: 'integer' },
            author: {
              name: { type: 'string' },
              email: { type: 'email' },
            },
          });
          this.raml.methods('books', 'get', {
            description: 'Get information about all books',
            responses: {
              200: { 'application/json': [{ name: 'one', author: { name: 'Bert' } }] },
              404: { 'application/json': { code: '120', message: 'Books was found' } },
            },
          });
          this.raml.methods('books', 'get', {
            description: 'Get information about all books',
            responses: {
              200: { 'application/json': [{ name: 'one', author: { name: 'Art' } }] },
              404: { 'application/json': { code: '120', message: 'Books not found' } },
            },
          });
        });

        afterEach(function () {
          this.raml = null;
        });

        it('check generated file', function (done) {
          this.raml.generate(function (err, ramlText) {
            (err instanceof Error).should.equal(false);
            ramlText.should.match(/#%RAML 1.0/);
            ramlText.should.match(/title: Testing/);
            ramlText.should.match(/baseUri: http:\/\/localhost:3000/);
            ramlText.should.match(/version: v1/);
            ramlText.should.match(/types:/);
            ramlText.should.match(/\/books/);
            ramlText.should.match(/numberOfPages/);
            ramlText.should.match(/Get information about all books/);
            ramlText.should.match(/200/);
            ramlText.should.match(/Art/);
            ramlText.should.match(/Books not found/);
            done();
          });
        });
      });

    describe('type and methods getters', function () {

        beforeEach(function () {
          this.raml = new Raml();
          this.raml.type('books', {
            name: { type: 'string', required: true },
            numberOfPages: { type: 'integer' },
            author: {
              name: { type: 'string' },
              email: { type: 'email' },
            },
          });
          this.raml.methods('books', 'get', {
            description: 'Get information about all books',
            responses: {
              200: { 'application/json': [{ name: 'one', author: { name: 'Art' } }] },
              404: { 'application/json': { code: '120', message: 'Books not found' } },
            },
          });
        });

        afterEach(function () {
          this.raml = null;
        });

        it('type testing', function (done) {
          this.raml.type().should.eql({
            books: {
              name: { type: 'string', required: true },
              numberOfPages: { type: 'integer' },
              author: { name: { type: 'string' }, email: { type: 'email' } },
            },
          });

          this.raml.type('books').should.eql({
            name: { type: 'string', required: true },
            numberOfPages: { type: 'integer' },
            author: { name: { type: 'string' }, email: { type: 'email' } },
          });

          done();
        });

        it('methods testing', function (done) {
          this.raml.methods().should.eql({
            books: {
              get: {
                description: 'Get information about all books',
                responses: {
                  200: { 'application/json': [{ name: 'one', author: { name: 'Art' } }] },
                  404: { 'application/json': { code: '120', message: 'Books not found' } },
                },
              },
            },
          });

          this.raml.methods('books').should.eql({
            get: {
              description: 'Get information about all books',
              responses: {
                200: { 'application/json': [{ name: 'one', author: { name: 'Art' } }] },
                404: { 'application/json': { code: '120', message: 'Books not found' } },
              },
            },
          });

          done();
        });

      });

    describe('empty type and methods getters', function () {

        beforeEach(function () {
          this.raml = new Raml();
        });

        afterEach(function () {
          this.raml = null;
        });

        it('type testing', function (done) {
          this.raml.type().should.eql({});
          (typeof this.raml.type('books') === 'undefined').should.equal(true);
          done();
        });

        it('methods testing', function (done) {
          this.raml.methods().should.eql({});
          (typeof this.raml.methods('books') === 'undefined').should.equal(true);
          done();
        });

      });

  });



test/express.js
'use strict';
var should = require('chai').should();
var express = require('express');
var httpMocks = require('node-mocks-http');
var Raml = require('../');

describe('testing express', function () {

  describe('check generated raml', function () {
    var app = { _router: { stack: [
      undefined,
      { },
      { route: { path: 'no_methods' } },
      { route: { methods: 'not_object' } },
      { route: { methods: { no_path: true } } },
      { route: { path: '/aaa', methods: { get: true } } },
      { route: { path: '/aaa', methods: { post: true } } },
      { route: { path: '/aaa/:idAaa', methods: { get: true } } },
      { route: { path: '/bbb', methods: { get: true } } },
    ], }, };
    app.use = function() {};
    app.get = function() {};

    beforeEach(function () {
      this.raml = new Raml({ express: app, storeResponses: true });
    });

    afterEach(function () {
      this.raml = null;
    });

    it('app is undefined', function (done) {
      this.raml.express({}, { send: function (data) {
          data.should.match(/^\/aaa:/m);
          data.should.match(/^\/bbb:/m);
          data.should.match(/\s\/{idAaa}:(.|\s)+\/bbb/);
          done();
        }, });
    });
  });

  [undefined, { _router: undefined }, { _router: { stack: undefined } }].forEach(function (app) {
    describe('check various empties app', function () {
      beforeEach(function () {
        if(app) app.use = function() {};
        if(app) app.get = function() {};
        this.raml = new Raml({ express: app, storeResponses: true });
      });

      afterEach(function () {
        this.raml = null;
      });

      it('app is undefined', function (done) {
        this.raml.express({}, { send: function (data) {
            data.should.match(/^#%RAML 1.0/m);
            data.should.match(/^title:/m);
            data.should.match(/^version:/m);
            (data.split('\n').length <= 6).should.equal(true);
            done();
          }, });
      });
    });
  });

  describe('express workflow', function () {

    beforeEach(function () {
      this.app = express()
      this.app.get('/b', function (req, res) { res.send('b1:') });
      this.app.post('/b', function (req, res) { res.send('b2:') });
      this.app.get('/b/:id', function (req, res) { res.send('b3:' + req.params.id) });
      this.app.delete('/b/:id', function (req, res) { res.send('b4:' + req.params.id) });

      this.raml = new Raml({express: this.app, storeResponses: true, guessAll: true});
      this.app.get('/api.raml', this.raml.express.bind(this.raml));
    });

    afterEach(function () {
      this.raml = null;
      this.app = null;
    });

    it('typical usage', function (done) {
      this.raml.express({}, { send: function (data) {
          data.should.match(/^\/b:/m);
          data.should.match(/^\s+\/{id}:/m);
          done();
        }, });
    });
  });

});

describe('with httpMocks no guessAll', function () {

    var app = { _router: { stack: [
      { route: { path: '/aaa', methods: { get: true } } },
      { route: { path: '/aaa', methods: { post: true } } },
      { route: { path: '/aaa/:id', methods: { get: true } } },
    ], }, };
    app.use = function() {};
    app.get = function() {};

    var raml = new Raml({ express: app, storeResponses: true });
    var request = {};
    var response = {};

    beforeEach(function(done) {
      response = httpMocks.createResponse({
        eventEmitter: require('events').EventEmitter
      });
      done();
    });

    it('post /aaa', function (done) {

      request = httpMocks.createRequest({
        method: 'POST',
        url: '/aaa',
        query: {
          name: 'foo'
        },
        headers: {
          'content-type' : 'application/json',
        },
        body: {
          name: 'foo'
        },
        route: {
          path: '/aaa',
          stack: [{
            method: 'post',
          }]
        }
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(201);
        response._headers['content-type'] = 'application/json';
        response.end('{"name":"foo"}');
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });

    it('post /aaa again', function (done) {

      request = httpMocks.createRequest({
        method: 'POST',
        url: '/aaa',
        query: {
          name: 'foo'
        },
        headers: {
          'content-type' : 'application/json',
        },
        body: {
          name: 'foo'
        },
        route: {
          path: '/aaa',
          stack: [{
            method: 'post',
          }]
        }
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(201);
        response._headers['content-type'] = 'application/json';
        response.end('{"error":"conflict"}');
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });
});

describe('with httpMocks', function () {

    var app = { _router: { stack: [
      { route: { path: '/aaa', methods: { get: true } } },
      { route: { path: '/aaa', methods: { post: true } } },
      { route: { path: '/aaa/:id', methods: { get: true } } },
    ], }, };
    app.use = function() {};
    app.get = function() {};

    var raml = new Raml({ express: app, storeResponses: true, guessAll: true });
    var request = {};
    var response = {};

    beforeEach(function(done) {
      response = httpMocks.createResponse({
        eventEmitter: require('events').EventEmitter
      });
      done();
    });

    it('get /aaa/1/bbb', function (done) {

      request = httpMocks.createRequest({
        method: 'GET',
        url: '/aaa/1/bbb',
        route: {
          path: '/aaa/:id/bbb',
          stack: [{
            method: 'get',
          }]
        }
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(200);
        response._headers['content-type'] = 'application/json';
        response.end('{"name":"foo"}');
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });

    it('post /aaa', function (done) {

      request = httpMocks.createRequest({
        method: 'POST',
        url: '/aaa',
        query: {
          name: 'foo'
        },
        headers: {
          'content-type' : 'application/json',
        },
        body: {
          name: 'foo'
        },
        route: {
          path: '/aaa',
          stack: [{
            method: 'post',
          }]
        }
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(201);
        response._headers['content-type'] = 'application/json';
        response.end('{"name":"foo"}');
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });

    it('post /aaa again', function (done) {

      request = httpMocks.createRequest({
        method: 'POST',
        url: '/aaa',
        query: {
          name: 'foo'
        },
        headers: {
          'content-type' : 'application/json',
        },
        body: {
          name: 'foo'
        },
        route: {
          path: '/aaa',
          stack: [{
            method: 'post',
          }]
        }
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(201);
        response._headers['content-type'] = 'application/json';
        response.end('{"error":"conflict"}');
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });

    it('post /aaa again', function (done) {

      request = httpMocks.createRequest({
        method: 'POST',
        url: '/aaa',
        query: {
          name: 'foo'
        },
        headers: {
          'content-type' : 'application/json',
        },
        body: {
          name: 'foo'
        },
        route: {
          path: '/aaa',
          stack: [{
            method: 'post',
          }]
        }
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(409);
        response._headers['content-type'] = 'text/other';
        response.end('{"error":"conflict"}');
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });

    it('post /aaa unknown content type', function (done) {

      request = httpMocks.createRequest({
        method: 'POST',
        url: '/aaa',
        query: {
          name: 'foo'
        },
        headers: {
          'content-type' : 'unknown',
        },
        body: {
          name: 'foo'
        },
        route: {
          path: '/aaa',
          stack: [{
            method: 'post',
          }]
        }
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(400);
        response._headers['content-type'] = 'unknown';
        response.end('{"name":"foo"}');
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });

    it('delete /aaa/123', function (done) {

      request = httpMocks.createRequest({
        method: 'DELETE',
        url: '/aaa/123',
        query: {
          name: 'foo'
        },
        route: {
          path: '/aaa/123',
          stack: [{
            method: 'delete',
          }]
        }
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(400);
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });

    it('delete /aaa/123 again', function (done) {

      request = httpMocks.createRequest({
        method: 'DELETE',
        url: '/aaa/123',
        query: {
          name: 'foo'
        },
        route: {
          path: '/aaa/123',
          stack: [{
            method: 'delete',
          }]
        }
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(400);
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });

    it('post /xxx', function (done) {

      request = httpMocks.createRequest({
        method: 'POST',
        url: '/xxx',
        query: {
          name: 'foo'
        },
        route: {
          path: '/xxx',
          stack: [{
            method: 'post',
          }]
        }
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(400);
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });

    it('no route', function (done) {
      request = httpMocks.createRequest({
        method: 'GET',
        url: '/aaaBBB',
        query: {
          name: 'foo'
        },
      });

      response.on('end', function () {
        // console.log(response._getData());
        response.status(400);
        done();
      });

      var _this = this;
      raml.storeResponses(request, response, function next(error) {
        raml.express(request, response, function next(error) {
        });
      });
    });

});



i.js
'use strict';

var fs = require('fs');

var Raml = require('./index');
var raml = new Raml({
  title: 'Testing Test',
  baseUri: 'http://localhost:3000',
  version: 'v1',
});

raml.type('books', {
  name: { type: 'string', required: true },
  numberOfPages: { type: 'integer' },
  author: {
    name: { type: 'string' },
    email: { type: 'email' },
  },
});

raml.type('user', {
  name: { type: 'string', required: true },
  email: { type: 'email' },
});

raml.generate(function (err, ramlText) {

  console.log(ramlText);

  fs.writeFile('test-raml2html.raml', ramlText, function (err) {
    console.log('===== raml2html parsing =====');
    var raml2html = require('raml2html');
    var configWithDefaultTheme = raml2html.getConfigForTheme();
    raml2html.render('test-raml2html.raml', configWithDefaultTheme).then(function (result) {
      console.log('test-raml2html passed');

    }, function (error) {

      console.log('Error raml2html parsing: ' + error);
    });
  });

});
/*
raml.generate(function(err, ramlText2) {

  console.log(ramlText2);

  fs.writeFile("test-raml2obj.raml", ramlText2, function(err) {
    console.log('===== raml2obj parsing =====');
    var raml2obj = require('raml2obj');
    raml2obj.parse('test-raml2obj.raml').then(function(ramlObj) {
      console.log('test-raml2obj passed');
    }, function(error) {
      console.log('Error raml2obj parsing: ' + error);
    });
  });

});
*/


index.js
/*!
 * create-raml
 * Copyright(c) 2017 ivanoff .$ curl -A cv ivanoff.org.ua
 * MIT Licensed
 */
'use strict';
var fs = require('fs');
var path = require('path');
var nunjucks = require('nunjucks');

exports = module.exports = function (options) {
  var typesData = {};
  var methodsData = {};

  if (undef(options)) options = {};
  var ramlFile = (options.version === 0.8) ? 'raml_0_8.nunjucks' : 'raml_1_0.nunjucks';
  var apiPath = options.path || '/api.raml';
  if (undef(options.templateFileName))
    options.templateFileName = path.resolve(__dirname, 'templates', ramlFile);

  var exp = {};

  exp.type = function (name, obj) {
    return (undef(name)) ? typesData
      : (undef(obj)) ? typesData[name]
      : typesData[name] = obj;
  };

  exp.methods = function (name, method, obj) {
    if (undef(name)) {
      return methodsData;
    } else if (undef(method)) {
      return methodsData[name];
    } else if (undef(obj)) {
      methodsData[name] = method;
    } else {
      if (undef(methodsData[name])) methodsData[name] = {};
      methodsData[name][method] = obj;
    }
  };

  exp.checkMethodsData = function () {
    if (options.express && Object.keys(methodsData).length === 0)
      methodsData = parseExpressData(options);
  };

  exp.express = function (req, res) {
    this.checkMethodsData();
    this.generate(function (err, data) {
      res.send(data);
    });
  };

  exp.storeResponses = function (req, res, next) {
    this.checkMethodsData();
    var resEnd = res.end;

    res.end = function (c) {
      resEnd.apply(res, arguments);
      if (options.storeResponses && req.route && req.route.path !== apiPath)
        methodsData[req.route.path] =
          updateMethodsData(methodsData[req.route.path], req, res, c, options);
    };

    next();
  };

  exp.generate = function (cb) {
    fs.readFile(options.templateFileName, function (err, template) {
      if (err) {
        cb(err);
      } else {
        var view = {
          options: options,
          types: typesData,
          url: Object.keys(methodsData).sort(),
          methods: getMethodsTree(methodsData),
          f: additional(),
        };
        cb(null, nunjucks.renderString(template.toString(), view));
      }
    });
  };

  if (options.express) {
    options.express.use(exp.storeResponses.bind(exp));
    options.express.get(apiPath, exp.express.bind(exp));
  }

  return exp;

};

function undef(v) {
  return typeof v === 'undefined';
}

function additional() {
  return {
    stringify: function (value, replacer, space, indent) {
      return JSON.stringify(value, replacer, space).replace(/^/gm, Array(indent).join(' '));
    },

    updatePath: function (path) {
      return path.replace(/^([^\/])/, '/$1').replace(/:([^/:-]+)/g, '{$1}');
    },

    regReplace: function (what, from, to, flags) {
      return what.replace(new RegExp(from, flags), to);
    },

    repeat: function (what, count) {
      return Array(count + 1).join(what);
    },
  };
}

function getMethodsTree(methodsData) {
  var result = {};
  var url = Object.keys(methodsData).sort();
  for (var i = url.length - 1; i >= 0; i--) {
    for (var j = i - 1; j >= 0; j--) {
      if (undef(result[url[i]]) && url[i].indexOf(url[j] + '/') == 0) {
        result[url[i]] = {
          shortLink: url[i].replace(url[j] + '/', ''),
          deep: url[j].replace(/^\//,'').split('/').length,
          data: methodsData[url[i]],
        };
      }
    }

    if (undef(result[url[i]])) {
      result[url[i]] = { shortLink: url[i], deep: 0, data: methodsData[url[i]] };
    }
  }

  return result;
}

function parseExpressData(options) {
  var app = options.express;
  if (undef(app) || undef(app._router)) return {};
  var s = app._router.stack;
  if (!Array.isArray(s)) return {};

  var result = {};
  for (var i = 0; i < s.length; i++) {
    if (undef(s[i])) continue;
    var r = s[i].route;
    if (undef(r) || undef(r.path) || undef(r.methods)) continue;
    if (undef(result[r.path])) result[r.path] = {};
    Object.keys(r.methods).forEach(function (method) {
      result[r.path][method] = {
        description: guessDescription( method, r.path, options.guessAll ),
      };
    });
  }

  return result;
}

function updateMethodsData(methodsPath, req, res, chunk, options) {
  var r = req.route;

  if (undef(methodsPath)) methodsPath = {};
  if (undef(methodsPath[r.stack[0].method]))
    methodsPath[r.stack[0].method] = {
      description: guessDescription( r.stack[0].method, r.path, options.guessAll ),
    };

  var m = methodsPath[r.stack[0].method];

  if (req.headers['content-type']) {
    var reqCType = req.headers['content-type'].match(/([^;]*(application|text)[^;]*)/);
    if (reqCType && reqCType[1]) {
      if (undef(m.body)) m.body = {};
      if (undef(m.body[reqCType[1]]))
        m.body[reqCType[1]] = {
          example: req.body,
        };
    }
  }

  if (res._headers['content-type']) {
    var resCType = res._headers['content-type'].match(/([^;]*(application|text)[^;]*)/);
    if (resCType && resCType[1]) {
      var result = chunk? chunk.toString() : '{}';
      if (resCType[1] === 'application/json') result = JSON.parse(result);
      if (undef(m.responses)) m.responses = {};
      if (undef(m.responses[res.statusCode])) m.responses[res.statusCode] = {};
      if (undef(m.responses[res.statusCode][resCType[1]]))
        m.responses[res.statusCode][resCType[1]] = result;
    }
  }

  return methodsPath;
}

function guessDescription( method, path, guess ) {
  var defaultResult = method + ' ' + path;
  if(!guess) return defaultResult;
  var responseText = {
    get : 'List all {names}',
    getOne: 'Get {name} with ID {id}',
    getSub: 'List of {subnames} that {name} ID {id} owns',
    post : 'Insert a new record in to {name} collection',
    postSub: 'Insert a new record in to {subname} collection that {name} ID {id} has',
    put : 'Replace {name} with ID {id}',
    patch : 'Modify {name} with ID {id}',
    delete : 'Delete {name} with ID {id}',
  };
  var result;
  var m = method.toLowerCase();
  var r;
  if( r = path.match(/^\/?([^\/]+?)s?\/:([^/:-]+)\/(([^\/]+?)s?)$/) ){
    m += 'Sub';
    result = responseText[m].replace('{name}',r[1]).replace('{id}','{'+r[2]+'}');
    result = result.replace('{subname}',r[4]).replace('{subnames}',r[3]);
  }
  if( r = path.match(/^\/?([^\/]+?)s?\/:([^/:-]+)$/) ){
    if(m === 'get') m = 'getOne';
    result = responseText[m].replace('{name}',r[1]).replace('{id}','{'+r[2]+'}');
  }
  if( r  = path.match(/^\/?(([^\/]+?)s?)$/) ){
    result = responseText[m].replace('{name}',r[2]).replace('{names}',r[1]);
  }
  return result || defaultResult;
}



i2.js
'use strict';
var fs = require('fs');

var Raml = require('./index');
var raml = new Raml({
  title: 'Testing',
  baseUri: 'http://localhost:3000',
  version: 'v1',
});

raml.type('books', {
  name: { type: 'string', required: true },
  numberOfPages: { type: 'integer' },
  author: {
    name: { type: 'string' },
    email: { type: 'email' },
  },
});

raml.methods('books', 'get', {
  description: 'Get information about all books',
  responses: {
    200: { 'application/json': [{ name: 'one', author: { name: 'Art' } }] },
    404: { 'application/json': { code: '120', message: 'Books not found' } },
  },
});

raml.methods('books', 'post', {
  description: 'Add new book',
  body: {
      'application/json': {
          type: 'books',
          example: { name: 'one', author: { name: 'Art' } },
        },
    },
  responses: {
    201: { 'application/json': { name: 'one', author: { name: 'Art' } } },
    400: { 'application/json': { code: '221', message: 'name not matched with string type' } },
  },
});

raml.methods('books/:bookId/readers', {
  get: {
    description: 'Get information about who has read {bookId} book',
    responses: {
      200: { 'application/json': [{ _id: 343, name: 'Mike' },{ _id: 324, name: 'John'}] },
      404: { 'application/json': { code: '121', message: 'Readers not found' } },
    },
  },
});

raml.methods('books/:bookId', {
  get: {
    description: 'Get information about {bookId} book',
    responses: {
      200: { 'application/json': { _id: 123, name: 'one', author: { name: 'Art' } } },
      404: { 'application/json': { code: '121', message: 'Book not found' } },
    },
  },
  delete: {
    description: 'Delete {bookId} book',
    responses: {
      200: { 'application/json': { ok: 1, _id: 123 } },
      400: { 'application/json': { code: '221', message: 'name not matched with string type' } },
      404: { 'application/json': { code: '121', message: 'Book not found' } },
    },
  },
});

raml.generate(function (err, ramlText) {

  console.log(ramlText);

  fs.writeFile('test-raml2html.raml', ramlText, function (err) {
    console.log('===== raml2html parsing =====');
    var raml2html = require('raml2html');
    var configWithDefaultTheme = raml2html.getConfigForTheme();
    raml2html.render('test-raml2html.raml', configWithDefaultTheme).then(function (result) {
      console.log('test-raml2html passed');
    }, function (error) {

      console.log('Error raml2html parsing: ' + error);
    });
  });

});

//var raml2 = new Raml();
//console.log(raml2.generate());
